python/reg_interface/scripts/gbt.py

```diff
@@ -15,6 +15,7 @@
         print('available commands:')
         print('  config <config_filename_txt>:   Configures the GBT with the given config file (must use the txt version of the config file, can be generated with the GBT programmer software)')
         print('  v3b-phase-scan <base_config_filename_txt>:   Configures the GBT with the given config file, and performs an elink phase scan while checking the VFAT communication for each phase')
+        print('  set-phase <phase>: Set the phase of a GBT')
```

Comment:
done

however, does it make more sense to make the second argument <elink> instead of <phase>?

the elinks are associated with multiple vfats (https://github.com/cms-gem-daq-project/reg_utils/blob/master/python/reg_interface/arm/gbt_utils.py#L16-L18), so this may deceive so the user into thinking that what he is setting will only affect on vfat

Reply:
I think you've misunderstood how this tool is used, look at your usage case:

```
gbt.py 0 2 set-phase 10 5
```

You're starting with OH0, selecting GBT2, then asking for VFAT10.  Within a GBT there is no sharing of e-links.  And if this is done **correctly** it will *only* affect the specific VFAT in question, and no others.

Reply:
here

https://github.com/cms-gem-daq-project/reg_utils/blob/master/python/reg_interface/arm/gbt_utils.py#L63-L67

it seems that the register that is being set depends only on the elink, not the GBT. Am I missing something?

Reply:
> it seems that the register that is being set depends only on the elink, not the GBT. Am I missing something?
 
Look at the outer loop.  It is only considering GBTX defined by `gbtSelect`.  Only e-links *that specific* GBT are considered.  There is a 1-to-1 mapping of VFAT to e-link in the hardware.

Reply:
from here

https://github.com/cms-gem-daq-project/reg_utils/blob/master/python/reg_interface/arm/gbt_utils.py#L16-L18

it seems that there are 10 elinks for 24 vfats

Reply:
e-link 0 on GBT0 is different from e-link 0 on GBT1 and is different from e-link 0 on GBT2...they are three *separate* chips each with their *own* unique set of hardware traces that are *not* shared.

Reply:
> ok, I was confused because the address that is being written to depends only the elink

It depends on the e-link of *that specific GBT*; they are *not* shared between GBTs (think how that would even work...if one GBT was controlling the two different VFATs on the same e-link then you'd *never* be able to independently set a different register on one of them in the current communication protocol....).

> but I guess because of these lines

Not only, you have *three* separate containers that specify three different sets of GBT mapping making use of three different sets of traces:

https://github.com/cms-gem-daq-project/reg_utils/blob/feb98eb9543028dd03cfbe0777acedd1c8a005ed/python/reg_interface/arm/gbt_utils.py#L16-L18
 
> Then I think the currently committed implementation is fine. Please let me know if you would like any other changes.
 
I mean did you test it...?

Reply:
To test it, I guess I should read back the phase-containing register in the GBT, but it is not clear to me how to do this. Can this be done with gem_reg?

Reply:
You could implement a `read-phase` command here and read it back after writing.

Reply:
> You're starting with OH0, selecting GBT2, then asking for VFAT10. Within a GBT there is no sharing of e-links. And if this is done correctly it will only affect the specific VFAT in question, and no others.

In the usage case that I understand this tool to attack, the `GBT` is not even the thing someone cares about, so why not simply use the VFAT as the lookup, and have the appropriate for both `[GBT,elink]` pair selected based on that.  Or do I miss something?

Reply:
Yes, that would logically make more sense, but would require major changes to the structure of gbt.py and gbt_utils.py and the command-line format. Currently, the user-requested GBT is used at the very beginning of the script:

https://github.com/cms-gem-daq-project/reg_utils/blob/master/python/reg_interface/scripts/gbt.py#L12-L22
https://github.com/cms-gem-daq-project/reg_utils/blob/master/python/reg_interface/arm/gbt_utils.py#L29-L33

Reply:
Possibly not for here, but the next thing to do with this tool would be to adapt the arguments/options like @bdorney did for `sca.py`

Reply:
> Possibly not for here, but the next thing to do with this tool would be to adapt the arguments/options like @bdorney did for sca.py
 
Yes that would be a very good move.  @AndrewLevin let's finish up this PR and then open a separate one.  In the meantime can you open a new issue for re-factoring the `gbt.py` tool in this manner.

Reply:
OK, here is the issue about refactoring gbt.py: https://github.com/cms-gem-daq-project/reg_utils/issues/38

Reply:
> there is currently no way to read a GBT register

That's fine, so a `read-phase` mode is not then very useful.  However you can still test that the phase was written correctly.  e.g. write a bad phase to a VFAT, it will no longer sync.  Then write a good phase and it will be able to sync again.

Reply:
ok, I have just performed a successful test of the script on eagle64 link 0:

http://cmsonline.cern.ch/cms-elog/1073245